### ISSUE TITLE
NaN cls loss on MPS. fg_mask.sum() is zero

### ISSUE THREAD

**Issue by kalkun** [ultralytics/ultralytics#9933]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

I've tried testing the same training command using `device=mps` vs `device=cpu`. When on MPS i get that the `fg_mask.sum()` is zero so the box and dfl loss are also 0. The  cls loss is `nan`. 

If i put a breakpoint in the assigner defined in `utils/tal.py` and compare the two commands, then it seems the pd_bboxes contains nan values on MPS which is not the case on CPU. See the screenshot below. I have the MPS command in the bottom left and the CPU in the bottom right: 
![image](https://github.com/ultralytics/ultralytics/assets/3143907/05639d1b-aeff-4792-b45e-c27faf639f12)

### Environment

```bash
❯ yolo check
Ultralytics YOLOv8.1.42 🚀 Python-3.9.19 torch-2.2.2 CPU (Apple M2)
Setup complete ✅ (8 CPUs, 16.0 GB RAM, 374.7/460.4 GB disk)

OS                  macOS-14.4.1-arm64-arm-64bit
Environment         Darwin
Python              3.9.19
Install             pip
RAM                 16.00 GB
CPU                 Apple M2
CUDA                None

matplotlib          ✅ 3.8.3>=3.3.0
opencv-python       ✅ 4.9.0.80>=4.6.0
pillow              ✅ 10.3.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.12.0>=1.4.1
torch               ✅ 2.2.2>=1.8.0
torchvision         ✅ 0.17.2>=0.9.0
tqdm                ✅ 4.66.2>=4.64.0
psutil              ✅ 5.9.8
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
pandas              ✅ 2.2.1>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
```

### Minimal Reproducible Example

```bash
yolo detect train model=yolov8n.pt epochs=10 imgsz=128 device=mps
```
vs 

```bash
yolo detect train model=yolov8n.pt epochs=10 imgsz=128
```

### Additional

I'm using a custom dataset and im not aware whether this persists on other non-custom datasets. 

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by kalkun** [comment]
Not sure whats going on exactly, but the scale tensor ends up having a `nan` value - though if i run the same expressions in the debugger, I dont get the expected results: 
![image](https://github.com/ultralytics/ultralytics/assets/3143907/38778932-8ffe-41f6-98ed-54a5d4d2ade5)

**Comment by kalkun** [comment]
I can resolve this by storing them one by one seemingly: 
![image](https://github.com/ultralytics/ultralytics/assets/3143907/74aab2e8-14dd-41d9-b4a8-2f4136286a63)

**Comment by kalkun** [comment]
I think this issue is related to: https://github.com/ultralytics/ultralytics/issues/9113

**Comment by glenn-jocher** [comment]
@kalkun could this issue also be related to https://github.com/ultralytics/ultralytics/issues/8555?

**Comment by glenn-jocher** [comment]
@kalkun for the MPS issue in #8555 user submitted PR https://github.com/ultralytics/ultralytics/pull/8639. Can you test this PR to see if the changes in it resolve your issue?

**Comment by kalkun** [comment]
> @kalkun for the MPS issue in #8555 user submitted PR #8639. Can you test this PR to see if the changes in it resolve your issue?

Thanks for pointing this out. Unfortunately it doesnt solve it for me entirely, though there are some observed changes from it. The `cls` loss is no longer nan but both `box` and `dfl` are still 0. Setting a breakpoint around the same place shows that the `fg_mask.sum()` is still 0 (which is not the case when on cpu). But the `imgsz` that i screenshot above too is well defined from that PR: 
![image](https://github.com/ultralytics/ultralytics/assets/3143907/12e85ab0-1539-48d5-ba68-9b646ea25564)

**Comment by glenn-jocher** [comment]
Thanks for giving that a try and for the feedback, @kalkun! 🙏 While it's good to hear that there has been some improvement regarding the `cls` loss, it's definitely concerning that the `box` and `dfl` losses remain at 0 and the `fg_mask.sum()` issue persists on MPS. This suggests there might be deeper compatibility issues at play specifically with MPS.

For further investigation, could you provide more details about your environment, especially the macOS version and the type of Mac Apple Silicon chip you're using? This could help us pinpoint any potential hardware-specific issues.

In the meantime, considering the peculiar behavior observed, I'll discuss this internally with our team to see if we've encountered similar issues or if there are any known workarounds specific to MPS devices. Any additional insights or peculiarities you notice would be helpful for us to better understand the situation.

Thanks again for your patience and assistance in debugging this! 👍

**Comment by kalkun** [comment]
Thanks @glenn-jocher and I've updated the description with the output of `yolo check`.

**Comment by glenn-jocher** [comment]
@kalkun thanks for updating with your environment details! It's helpful for diagnosing the issue. ✨ If specific parts of your setup are unusual or customized, considering updating those details as well might provide extra clues.

If you haven't yet, running a simple sanity check script to ensure MPS is functioning correctly outside of our context might also be insightful. Here’s a small snippet you can run:

```python
import torch
print(torch.rand(3,3).to('mps'))  # just a simple operation on MPS
```

This can help determine if the issue is truly with the MPS integration or something deeper in the codebase. If this script runs without any errors, the MPS setup is likely correct. Keep us posted!
